# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 12)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 12:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .
the lime is her least liked fruit , but the banana is my least liked .
he s

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_id_text = [[source_vocab_to_int[word] for word in sentence.split()] 
                      for sentence in source_text.split('\n')]
    
    target_id_text = [[target_vocab_to_int[word] for word in sentence.split()]
                      + [target_vocab_to_int['<EOS>']]
                      for sentence in target_text.split('\n')]
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


D:\Anaconda3\envs\py35\lib\site-packages\ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_ = tf.placeholder(tf.int32, [None, None], name = "input")
    targets = tf.placeholder(tf.int32, [None, None], name = "targets")
    learning_rate = tf.placeholder(tf.float32, None, name = "learning_rate")
    keep_probability = tf.placeholder(tf.float32, None, name = "keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, [None], name = "target_sequence_length")
    
    #axis = None -> Rank 0
    max_target_length = tf.reduce_max(input_tensor = target_sequence_length, axis = None, name = "max_target_len")
    
    source_sequence_length = tf.placeholder(tf.int32, [None], name = "source_sequence_length")
    
    return input_, targets, learning_rate, keep_probability, target_sequence_length, max_target_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    
    #https://www.tensorflow.org/api_docs/python/tf/strided_slice
    ending = tf.strided_slice(input_ = target_data, 
                              begin = [0,0], 
                              end = [batch_size, -1], 
                              strides = [1, 1])
       
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    
    # encoder embedding
    # https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence
    embed_input = tf.contrib.layers.embed_sequence(ids = rnn_inputs, 
                                                   vocab_size = source_vocab_size, 
                                                   embed_dim = encoding_embedding_size)
    
    # RNN cell
    def lstm_cell():
        #lstm = tf.contrib.rnn.LSTMCell(rnn_size, initializer = tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        lstm = tf.contrib.rnn.LSTMCell(rnn_size)
        lstm_dropout = tf.contrib.rnn.DropoutWrapper(lstm, state_keep_prob = keep_prob)
        return lstm_dropout
    
    stacked_lstm = tf.contrib.rnn.MultiRNNCell(
        [lstm_cell() for _ in range(num_layers)])
    
    
        
    rnn_output, rnn_state = tf.nn.dynamic_rnn(stacked_lstm, 
                                              embed_input, 
                                              sequence_length = source_sequence_length,
                                              dtype = tf.float32)
    
    return rnn_output, rnn_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    
    # Helper for the training process
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs = dec_embed_input, 
                                                        sequence_length = target_sequence_length)

    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(cell = dec_cell, 
                                                       helper = training_helper, 
                                                       initial_state = encoder_state, 
                                                       output_layer = output_layer)

    # Perform dynamic decoding using the decoder
    training_decoder_output = \
        tf.contrib.seq2seq.dynamic_decode(decoder = training_decoder, 
                                          impute_finished = True,
                                          maximum_iterations = max_summary_length)[0]

    # Dropout
    #training_decoder_output = tf.nn.dropout(training_decoder_output, keep_prob)
   
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function    
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name="start_tokens")
    
    # Helper for the inference process
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(embedding = dec_embeddings, 
                                                                start_tokens = start_tokens, 
                                                                end_token = end_of_sequence_id)
    
    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(cell = dec_cell, 
                                                        helper = inference_helper, 
                                                        initial_state = encoder_state, 
                                                        output_layer = output_layer)
    
    # Perform dynamic decoding using the decoder
    inference_decoder_output = \
        tf.contrib.seq2seq.dynamic_decode(decoder = inference_decoder, 
                                          impute_finished = True, 
                                          maximum_iterations = max_target_sequence_length)[0]
    
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    # 1. Decoder Embedding
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size, 
                                           initializer = tf.random_uniform_initializer(-0.1, 0.1, seed = 2))
        return dec_cell
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    # 3. Dense layer to translate the decoder's output at each time
    output_layer = Dense(target_vocab_size, 
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    # 4. Set up a training decoder and an inference decoder
    with tf.variable_scope("decode"):
        training_decoder_output = \
            decoding_layer_train(encoder_state = encoder_state, 
                                 dec_cell = dec_cell, 
                                 dec_embed_input = dec_embed_input, 
                                 target_sequence_length = target_sequence_length, 
                                 max_summary_length = max_target_sequence_length, 
                                 output_layer = output_layer, 
                                 keep_prob = keep_prob)
               
    # 5. Inference Decoder
    with tf.variable_scope("decode", reuse = True):
        inference_decoder_output = \
            decoding_layer_infer(encoder_state = encoder_state, 
                                 dec_cell = dec_cell, 
                                 dec_embeddings = dec_embeddings, 
                                 start_of_sequence_id = target_vocab_to_int['<GO>'], 
                                 end_of_sequence_id = target_vocab_to_int['<EOS>'], 
                                 max_target_sequence_length = max_target_sequence_length, 
                                 vocab_size = target_vocab_size, 
                                 output_layer = output_layer, 
                                 batch_size = batch_size, 
                                 keep_prob = keep_prob)
    
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    # pass the input data through the encoder
    _, enc_state = encoding_layer(rnn_inputs = input_data, 
                                  rnn_size = rnn_size, 
                                  num_layers = num_layers, 
                                  keep_prob = keep_prob, 
                                  source_sequence_length = source_sequence_length, 
                                  source_vocab_size = source_vocab_size, 
                                  encoding_embedding_size = enc_embedding_size)
    
    # prepare the target sequences
    dec_input = process_decoder_input(target_data = target_data, 
                                      target_vocab_to_int = target_vocab_to_int, 
                                      batch_size = batch_size)
    
    # pass encoder and decoder input to the decoder
    training_decoder_output, inference_decoder_output = \
        decoding_layer(dec_input = dec_input, 
                       encoder_state = enc_state, 
                       target_sequence_length = target_sequence_length, 
                       max_target_sequence_length = max_target_sentence_length, 
                       rnn_size = rnn_size, 
                       num_layers = num_layers, 
                       target_vocab_to_int = target_vocab_to_int, 
                       target_vocab_size = target_vocab_size, 
                       batch_size = batch_size, 
                       keep_prob = keep_prob, 
                       decoding_embedding_size = dec_embedding_size)
    
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 15
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.8
display_step = 10

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/538 - Train Accuracy: 0.3125, Validation Accuracy: 0.3949, Loss: 3.3830
Epoch   0 Batch   20/538 - Train Accuracy: 0.3640, Validation Accuracy: 0.4015, Loss: 2.6429
Epoch   0 Batch   30/538 - Train Accuracy: 0.3592, Validation Accuracy: 0.4146, Loss: 2.5186
Epoch   0 Batch   40/538 - Train Accuracy: 0.4938, Validation Accuracy: 0.4982, Loss: 2.2071
Epoch   0 Batch   50/538 - Train Accuracy: 0.4594, Validation Accuracy: 0.4979, Loss: 2.1938
Epoch   0 Batch   60/538 - Train Accuracy: 0.4395, Validation Accuracy: 0.4895, Loss: 2.1318
Epoch   0 Batch   70/538 - Train Accuracy: 0.4702, Validation Accuracy: 0.4904, Loss: 1.9351
Epoch   0 Batch   80/538 - Train Accuracy: 0.4592, Validation Accuracy: 0.5053, Loss: 1.9091
Epoch   0 Batch   90/538 - Train Accuracy: 0.4820, Validation Accuracy: 0.5146, Loss: 1.6996
Epoch   0 Batch  100/538 - Train Accuracy: 0.4850, Validation Accuracy: 0.5281, Loss: 1.5864
Epoch   0 Batch  110/538 - Train Accuracy: 0.4809, Validation Accuracy

Epoch   1 Batch  370/538 - Train Accuracy: 0.6820, Validation Accuracy: 0.7152, Loss: 0.4043
Epoch   1 Batch  380/538 - Train Accuracy: 0.6715, Validation Accuracy: 0.7214, Loss: 0.3730
Epoch   1 Batch  390/538 - Train Accuracy: 0.7189, Validation Accuracy: 0.7015, Loss: 0.3717
Epoch   1 Batch  400/538 - Train Accuracy: 0.6940, Validation Accuracy: 0.7072, Loss: 0.3635
Epoch   1 Batch  410/538 - Train Accuracy: 0.7041, Validation Accuracy: 0.7198, Loss: 0.3708
Epoch   1 Batch  420/538 - Train Accuracy: 0.7004, Validation Accuracy: 0.6955, Loss: 0.3644
Epoch   1 Batch  430/538 - Train Accuracy: 0.7082, Validation Accuracy: 0.7113, Loss: 0.3604
Epoch   1 Batch  440/538 - Train Accuracy: 0.6945, Validation Accuracy: 0.7262, Loss: 0.3604
Epoch   1 Batch  450/538 - Train Accuracy: 0.7214, Validation Accuracy: 0.7182, Loss: 0.3466
Epoch   1 Batch  460/538 - Train Accuracy: 0.6897, Validation Accuracy: 0.7306, Loss: 0.3289
Epoch   1 Batch  470/538 - Train Accuracy: 0.7197, Validation Accuracy

Epoch   3 Batch  200/538 - Train Accuracy: 0.8826, Validation Accuracy: 0.8720, Loss: 0.1253
Epoch   3 Batch  210/538 - Train Accuracy: 0.8387, Validation Accuracy: 0.8597, Loss: 0.1444
Epoch   3 Batch  220/538 - Train Accuracy: 0.8637, Validation Accuracy: 0.8517, Loss: 0.1291
Epoch   3 Batch  230/538 - Train Accuracy: 0.8697, Validation Accuracy: 0.8695, Loss: 0.1375
Epoch   3 Batch  240/538 - Train Accuracy: 0.8885, Validation Accuracy: 0.8716, Loss: 0.1208
Epoch   3 Batch  250/538 - Train Accuracy: 0.8818, Validation Accuracy: 0.8448, Loss: 0.1355
Epoch   3 Batch  260/538 - Train Accuracy: 0.8635, Validation Accuracy: 0.8704, Loss: 0.1291
Epoch   3 Batch  270/538 - Train Accuracy: 0.8746, Validation Accuracy: 0.8871, Loss: 0.1265
Epoch   3 Batch  280/538 - Train Accuracy: 0.8992, Validation Accuracy: 0.8833, Loss: 0.1162
Epoch   3 Batch  290/538 - Train Accuracy: 0.9018, Validation Accuracy: 0.8667, Loss: 0.1144
Epoch   3 Batch  300/538 - Train Accuracy: 0.8891, Validation Accuracy

Epoch   5 Batch   30/538 - Train Accuracy: 0.9484, Validation Accuracy: 0.9409, Loss: 0.0446
Epoch   5 Batch   40/538 - Train Accuracy: 0.9608, Validation Accuracy: 0.9585, Loss: 0.0323
Epoch   5 Batch   50/538 - Train Accuracy: 0.9572, Validation Accuracy: 0.9473, Loss: 0.0356
Epoch   5 Batch   60/538 - Train Accuracy: 0.9535, Validation Accuracy: 0.9457, Loss: 0.0447
Epoch   5 Batch   70/538 - Train Accuracy: 0.9522, Validation Accuracy: 0.9437, Loss: 0.0366
Epoch   5 Batch   80/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9622, Loss: 0.0450
Epoch   5 Batch   90/538 - Train Accuracy: 0.9611, Validation Accuracy: 0.9442, Loss: 0.0466
Epoch   5 Batch  100/538 - Train Accuracy: 0.9619, Validation Accuracy: 0.9586, Loss: 0.0358
Epoch   5 Batch  110/538 - Train Accuracy: 0.9588, Validation Accuracy: 0.9371, Loss: 0.0461
Epoch   5 Batch  120/538 - Train Accuracy: 0.9711, Validation Accuracy: 0.9476, Loss: 0.0337
Epoch   5 Batch  130/538 - Train Accuracy: 0.9608, Validation Accuracy

Epoch   6 Batch  390/538 - Train Accuracy: 0.9669, Validation Accuracy: 0.9640, Loss: 0.0228
Epoch   6 Batch  400/538 - Train Accuracy: 0.9671, Validation Accuracy: 0.9604, Loss: 0.0250
Epoch   6 Batch  410/538 - Train Accuracy: 0.9785, Validation Accuracy: 0.9581, Loss: 0.0209
Epoch   6 Batch  420/538 - Train Accuracy: 0.9691, Validation Accuracy: 0.9553, Loss: 0.0285
Epoch   6 Batch  430/538 - Train Accuracy: 0.9641, Validation Accuracy: 0.9698, Loss: 0.0282
Epoch   6 Batch  440/538 - Train Accuracy: 0.9762, Validation Accuracy: 0.9677, Loss: 0.0267
Epoch   6 Batch  450/538 - Train Accuracy: 0.9528, Validation Accuracy: 0.9647, Loss: 0.0334
Epoch   6 Batch  460/538 - Train Accuracy: 0.9641, Validation Accuracy: 0.9629, Loss: 0.0293
Epoch   6 Batch  470/538 - Train Accuracy: 0.9663, Validation Accuracy: 0.9659, Loss: 0.0247
Epoch   6 Batch  480/538 - Train Accuracy: 0.9786, Validation Accuracy: 0.9673, Loss: 0.0213
Epoch   6 Batch  490/538 - Train Accuracy: 0.9706, Validation Accuracy

Epoch   8 Batch  220/538 - Train Accuracy: 0.9676, Validation Accuracy: 0.9656, Loss: 0.0201
Epoch   8 Batch  230/538 - Train Accuracy: 0.9729, Validation Accuracy: 0.9631, Loss: 0.0208
Epoch   8 Batch  240/538 - Train Accuracy: 0.9719, Validation Accuracy: 0.9656, Loss: 0.0193
Epoch   8 Batch  250/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9636, Loss: 0.0190
Epoch   8 Batch  260/538 - Train Accuracy: 0.9594, Validation Accuracy: 0.9691, Loss: 0.0239
Epoch   8 Batch  270/538 - Train Accuracy: 0.9834, Validation Accuracy: 0.9663, Loss: 0.0211
Epoch   8 Batch  280/538 - Train Accuracy: 0.9756, Validation Accuracy: 0.9721, Loss: 0.0157
Epoch   8 Batch  290/538 - Train Accuracy: 0.9867, Validation Accuracy: 0.9773, Loss: 0.0151
Epoch   8 Batch  300/538 - Train Accuracy: 0.9760, Validation Accuracy: 0.9826, Loss: 0.0190
Epoch   8 Batch  310/538 - Train Accuracy: 0.9754, Validation Accuracy: 0.9728, Loss: 0.0223
Epoch   8 Batch  320/538 - Train Accuracy: 0.9840, Validation Accuracy

Epoch  10 Batch   50/538 - Train Accuracy: 0.9703, Validation Accuracy: 0.9826, Loss: 0.0213
Epoch  10 Batch   60/538 - Train Accuracy: 0.9855, Validation Accuracy: 0.9808, Loss: 0.0161
Epoch  10 Batch   70/538 - Train Accuracy: 0.9879, Validation Accuracy: 0.9730, Loss: 0.0164
Epoch  10 Batch   80/538 - Train Accuracy: 0.9807, Validation Accuracy: 0.9801, Loss: 0.0153
Epoch  10 Batch   90/538 - Train Accuracy: 0.9730, Validation Accuracy: 0.9750, Loss: 0.0179
Epoch  10 Batch  100/538 - Train Accuracy: 0.9824, Validation Accuracy: 0.9750, Loss: 0.0157
Epoch  10 Batch  110/538 - Train Accuracy: 0.9803, Validation Accuracy: 0.9705, Loss: 0.0176
Epoch  10 Batch  120/538 - Train Accuracy: 0.9891, Validation Accuracy: 0.9696, Loss: 0.0163
Epoch  10 Batch  130/538 - Train Accuracy: 0.9792, Validation Accuracy: 0.9799, Loss: 0.0190
Epoch  10 Batch  140/538 - Train Accuracy: 0.9705, Validation Accuracy: 0.9751, Loss: 0.0223
Epoch  10 Batch  150/538 - Train Accuracy: 0.9758, Validation Accuracy

Epoch  11 Batch  410/538 - Train Accuracy: 0.9900, Validation Accuracy: 0.9727, Loss: 0.0101
Epoch  11 Batch  420/538 - Train Accuracy: 0.9799, Validation Accuracy: 0.9743, Loss: 0.0177
Epoch  11 Batch  430/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9698, Loss: 0.0134
Epoch  11 Batch  440/538 - Train Accuracy: 0.9816, Validation Accuracy: 0.9739, Loss: 0.0149
Epoch  11 Batch  450/538 - Train Accuracy: 0.9732, Validation Accuracy: 0.9739, Loss: 0.0168
Epoch  11 Batch  460/538 - Train Accuracy: 0.9825, Validation Accuracy: 0.9648, Loss: 0.0126
Epoch  11 Batch  470/538 - Train Accuracy: 0.9821, Validation Accuracy: 0.9718, Loss: 0.0143
Epoch  11 Batch  480/538 - Train Accuracy: 0.9874, Validation Accuracy: 0.9755, Loss: 0.0101
Epoch  11 Batch  490/538 - Train Accuracy: 0.9907, Validation Accuracy: 0.9727, Loss: 0.0126
Epoch  11 Batch  500/538 - Train Accuracy: 0.9902, Validation Accuracy: 0.9844, Loss: 0.0119
Epoch  11 Batch  510/538 - Train Accuracy: 0.9887, Validation Accuracy

Epoch  13 Batch  240/538 - Train Accuracy: 0.9832, Validation Accuracy: 0.9748, Loss: 0.0188
Epoch  13 Batch  250/538 - Train Accuracy: 0.9885, Validation Accuracy: 0.9803, Loss: 0.0176
Epoch  13 Batch  260/538 - Train Accuracy: 0.9760, Validation Accuracy: 0.9714, Loss: 0.0173
Epoch  13 Batch  270/538 - Train Accuracy: 0.9836, Validation Accuracy: 0.9755, Loss: 0.0140
Epoch  13 Batch  280/538 - Train Accuracy: 0.9916, Validation Accuracy: 0.9755, Loss: 0.0115
Epoch  13 Batch  290/538 - Train Accuracy: 0.9941, Validation Accuracy: 0.9783, Loss: 0.0075
Epoch  13 Batch  300/538 - Train Accuracy: 0.9827, Validation Accuracy: 0.9751, Loss: 0.0122
Epoch  13 Batch  310/538 - Train Accuracy: 0.9834, Validation Accuracy: 0.9812, Loss: 0.0133
Epoch  13 Batch  320/538 - Train Accuracy: 0.9821, Validation Accuracy: 0.9741, Loss: 0.0112
Epoch  13 Batch  330/538 - Train Accuracy: 0.9914, Validation Accuracy: 0.9711, Loss: 0.0081
Epoch  13 Batch  340/538 - Train Accuracy: 0.9863, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    list_wordid = [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.lower().split()]
    return list_wordid


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [62, 172, 99, 122, 199, 70, 81]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [90, 214, 80, 39, 244, 150, 37, 119, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.